File: androidchallenge1/src/main/java/class28/androidchallenge01/ui/restaurants/AdapterRestaurantsList.java
package class28.androidchallenge01.ui.restaurants;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import class28.androidchallenge01.R;
import class28.androidchallenge01.model.Restaurant;
import class28.androidchallenge01.ui.plates.ActRestaurantPlates;

import static class28.androidchallenge01.ui.restaurants.FragRestaurants.CHOSEN_RESTAURANT;

class AdapterRestaurantsList extends RecyclerView.Adapter<RestaurantsListViewHolder> {

    private List<Restaurant> restaurantList;

    AdapterRestaurantsList(ArrayList<Restaurant> restaurantList, OnRestaurantItemClickCallback listener) {
        this.restaurantList = restaurantList;
    }

    @NonNull
    @Override
    public RestaurantsListViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new RestaurantsListViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_restaurant_list, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull RestaurantsListViewHolder holder, final int position) {
        final Restaurant restaurantItem = restaurantList.get(position);

        holder.restaurantImage.setImageResource(restaurantItem.getRestaurantImage());
        holder.restaurantName.setText(restaurantItem.getName());
        holder.restaurantAddress.setText(restaurantItem.getAddress());
        holder.restaurantClosesAt.setText(holder.itemView.getContext().getString(R.string.close_at_hhmm, restaurantItem.getCloseAt()));

        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Context activity = v.getContext();
//                Para usar interface, dessa forma seria a melhor
//                ((OnRestaurantItemClickCallback) activity).onRestaurantItemClickCallback(position);
                activity.startActivity(new Intent(activity, ActRestaurantPlates.class)
                        .putExtra(CHOSEN_RESTAURANT, restaurantItem));
            }
        });
    }

    @Override
    public int getItemCount() {
        return restaurantList.size();
    }
}

File: androidchallenge1kotlin/src/main/java/class28/androidchallenge01kotlin/ui/restaurants/AdapterRestaurantsList.kt
package class28.androidchallenge01kotlin.ui.restaurants

import android.content.Intent
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import class28.androidchallenge01kotlin.R
import class28.androidchallenge01kotlin.model.Restaurant
import class28.androidchallenge01kotlin.ui.plates.ActRestaurantPlates
import class28.androidchallenge01kotlin.ui.restaurants.FragRestaurants.Companion.CHOSEN_RESTAURANT

class AdapterRestaurantsList(
    private val restaurantList: List<Restaurant>,
    listener: OnRestaurantItemClickCallback
) : RecyclerView.Adapter<RestaurantsListViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = RestaurantsListViewHolder(
        LayoutInflater.from(parent.context).inflate(R.layout.item_restaurant_list, parent, false)
    )

    override fun getItemCount(): Int = restaurantList.size

    override fun onBindViewHolder(holder: RestaurantsListViewHolder, position: Int) {
        restaurantList[position].run {

            holder.apply {
                restaurantImage.setImageResource(restaurantImageRes)
                restaurantName.text = name
                restaurantAddress.text = address
                restaurantClosesAt.text = itemView.context.getString(R.string.close_at_hhmm, closeAt)

                itemView.setOnClickListener { v ->
                    val activity = v.context
                    //Para usar interface, dessa forma seria a melhor
//                    ((OnRestaurantItemClickCallback) activity).onRestaurantItemClickCallback(position);
                    activity.startActivity(Intent(activity, ActRestaurantPlates::class.java).putExtra(CHOSEN_RESTAURANT, this@run))
                }
            }
        }
    }
}